Tweaks for CMake 4 and GitHub CI

v-dobrev opened the pull request:
Update the CMake minimum version to 3.12.0 and add policy version.

Add a patch for CMake 4 for the "metis-mfem" vcpkg in `config/vcpkg`.

<!--GHEX{"author":"v-dobrev","assignment":"2025-04-02T00:17:37","editor":"tzanio","id":4774,"reviewers":["pazner","helloworld922","jandrej"],"merge":"2025-04-02T21:53:30","approval":"2025-04-02T20:26:15"}XEHG-->
<!--GHEXTABLE-->
 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#4774](https://github.com/mfem/mfem/pull/4774) | @v-dobrev | @tzanio | @pazner + @helloworld922 + @jandrej | 4/1/25 | 4/2/25 | 4/2/25 | |
<!--ELBATXEHG-->
<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

tzanio commented:
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

jandrej reviewed (commented):
LGTM

helloworld922 commented:
I'm running into some problems with trying to build this with CUDA. It looks like updating the CMake version now sets `CMAKE_CUDA_ARCHITECTURES` to a default value when CUDA is enabled, so a user-set CUDA_ARCH doesn't work. Should we remove the `CUDA_ARCH` option entirely and just rely on `CMAKE_CUDA_ARCHITECTURES`?

helloworld922 commented:
Added a fix to #4775 which removes CUDA_ARCH in favor of CMAKE_CUDA_ARCHITECTURES, emulating the behavior for CMake < 3.18.

helloworld922 commented:
Fix now also removes HIP_ARCH in favor of CMAKE_HIP_ARCHITECTURES.

v-dobrev commented:
I'd like to keep things in the CMake build system backward-compatible, so I prefer to keep both `CUDA_ARCH` and `HIP_ARCH`.

According to https://cmake.org/cmake/help/latest/variable/CMAKE_CUDA_ARCHITECTURES.html, `CMAKE_CUDA_ARCHITECTURES` may be initialized when the `CUDA` language is enabled, so we should be able to pre-initialize `CMAKE_CUDA_ARCHITECTURES` (if not already set by the user) before we call `enable_language(CUDA)`.

I'll make this change here.

v-dobrev commented:
It looks like the CMake builds with CUDA on Lassen are broken even in `master`, at least with older versions of CMake like 3.20. The CMake version that worked for me is 3.29.

Here are two of the issues:
* We use `CUDAToolkit` which requires CMake >= 3.17
* We use commands like 
    ```cmake
    get_target_property(CUBLAS_LIBRARIES CUDA::cublas LOCATION)
    ```
    which do not work with older CMake versions like 3.20.

I'll push some changes here to address these issues in a bit.

v-dobrev commented:
Okay, I think I fixed the CUDA issues with CMake. I also fixed the warnings from CUDA 11.1.1 -- they can be seen in the output from the Gitlab CI on Lassen, e.g. here: https://lc.llnl.gov/gitlab/mfem/mfem-autotest/-/jobs/2533430/raw (search for "warning").

@helloworld922, can you try this with your setup on Lassen to check if I missed something? Thanks!

adam-sim-dev commented on config/cmake/modules/MfemCmakeUtilities.cmake:
```diff
@@ -157,6 +150,21 @@ macro(add_mfem_miniapp MFEM_EXE_NAME)
+# successfully. Example usage:
+#   mfem_culib_set_libraries(CUBLAS cublas)
+macro(mfem_culib_set_libraries _CULIB _culib)
+  # The following command do not work with older CMake versions, e.g. 3.20:
```
does not

adam-sim-dev replied:
3.20 confused me. CMake minimal version 3.12 is required.

v-dobrev replied:
Typo fixed.

I'm not sure what is confusing about the statement: CMake 3.20 is a supported older version of CMake. If you have a suggestion how to improve it, let me know.

tzanio commented:
Thanks @v-dobrev!

Should we update `INSTALL` and mention this in `CHANGELOG`?

v-dobrev commented:
> > After searching the code, I found that "CUDA" is used in most places in the comments.
> 
> I found "Cuda" and "Hip" used in just a handful of comments, I vote for changing these to "CUDA" and "HIP" in this or the release PR... Any objections?

No objections, I just copy-pasted (and edited a bit) an existing comment that was already using the spelling "Cuda".

tzanio commented:
Thanks @v-dobrev!

I can handle the "Cuda/Hip -> CUDA/HIP" renaming after this is merged in the release branch (unless @helloworld922 objects).

tzanio commented:
Can we get additional approvals or feedback please? I'd like to merge this in `next`, `master` and `mfem-4.8-dev` soon.

tzanio commented:
Merged in `next` for testing...

adam-sim-dev commented:
A few days ago, I used CMake 3.31.6. mfem_LIB_DEPENDS is defined in CMakeCache.txt. Now it is missing. I don't know why.
